### PR TITLE
libp2p - refactoring around Capability

### PR DESCRIPTION
**Pull request by gumb0**
- Got rid of `Session` being friend of `Capability`
- Got rid of raw pointer to `HostCapabilityFace` as a member of `Capability`
- Removed `Capability::disconnect()` (`SessionFace::disconnect` is enough for this)

A couple of small refactoring more will go to the next PR.

### CODE REVIEW

**Comment by gumb0** [comment]
Continuation: https://github.com/ethereum/aleth/pull/5213

File: libethereum/EthereumPeer.h [hunk 1]
```diff
@@ -40,39 +40,39 @@ namespace eth
...
 
-	virtual void onPeerAborting() = 0;
+    virtual void onPeerAborting() = 0;
 };
 
 class EthereumHostDataFace
 {
 public:
-	virtual ~EthereumHostDataFace() {}
+    virtual ~EthereumHostDataFace() {}
```
Comment (chfast) [inline libethereum/EthereumPeer.h]: `= default`.

File: libethereum/EthereumPeer.h [hunk 2]
```diff
@@ -82,118 +82,119 @@ class EthereumHostDataFace
...
-	/// Basic constructor.
-	EthereumPeer(std::shared_ptr<p2p::SessionFace> _s, p2p::HostCapabilityFace* _h, unsigned _i, p2p::CapDesc const& _cap);
+    /// Basic constructor.
+    EthereumPeer(std::shared_ptr<p2p::SessionFace> _s, std::string const& _name,
+        unsigned _messageCount, unsigned _offset, p2p::CapDesc const& _cap);
 
-	/// Basic destructor.
-	virtual ~EthereumPeer();
+    /// Basic destructor.
+    virtual ~EthereumPeer();
```
Comment (chfast) [inline libethereum/EthereumPeer.h]: `override` instead of `virtual`.

File: libethereum/WarpPeerCapability.cpp [hunk 3]
```diff
@@ -27,8 +27,9 @@ namespace dev
 namespace eth
 {
 WarpPeerCapability::WarpPeerCapability(std::shared_ptr<p2p::SessionFace> _s,
-    p2p::HostCapabilityFace* _h, unsigned _i, p2p::CapDesc const& /* _cap */)
-  : Capability(_s, _h, _i)
+    std::string const& _name, unsigned _messageCount, unsigned _offset,
+    p2p::CapDesc const& /* _cap */)
+  : Capability(_s, _name, _messageCount, _offset)
```
Comment (chfast) [inline libethereum/WarpPeerCapability.cpp]: `std::move(_s)`.

File: libp2p/Capability.cpp [hunk 4]
```diff
@@ -28,23 +28,16 @@ using namespace std;
 using namespace dev;
 using namespace dev::p2p;
 
-Capability::Capability(std::shared_ptr<SessionFace> _s, HostCapabilityFace* _h, unsigned _idOffset):
-    m_session(_s), m_hostCap(_h), m_idOffset(_idOffset)
+Capability::Capability(std::shared_ptr<SessionFace> _s, string const& _name, unsigned _messageCount,
+    unsigned _idOffset)
+  : m_session(_s), m_name(_name), m_messageCount(_messageCount), m_idOffset(_idOffset)
```
Comment (chfast) [inline libp2p/Capability.cpp]: `std::move(_s)`.

File: libp2p/Capability.h [hunk 5]
```diff
@@ -33,28 +33,32 @@ class ReputationManager;
 
 class Capability: public std::enable_shared_from_this<Capability>
 {
-    friend class Session;
-
 public:
-    Capability(std::shared_ptr<SessionFace> _s, HostCapabilityFace* _h, unsigned _idOffset);
+    Capability(std::shared_ptr<SessionFace> _s, std::string const& _name, unsigned _messageCount,
+        unsigned _idOffset);
     virtual ~Capability() {}
```
Comment (chfast) [inline libp2p/Capability.h]: `= default`.

File: libp2p/HostCapability.h [hunk 6]
```diff
@@ -64,7 +64,7 @@ class HostCapability: public HostCapabilityFace
     std::shared_ptr<Capability> newPeerCapability(
         std::shared_ptr<SessionFace> const& _s, unsigned _idOffset, CapDesc const& _cap) override
     {
-        auto p = std::make_shared<PeerCap>(_s, this, _idOffset, _cap);
+        auto p = std::make_shared<PeerCap>(_s, name(), messageCount(), _idOffset, _cap);
```
Comment (chfast) [inline libp2p/HostCapability.h]: Why the capability `PeerCap` must own the session?

Comment (gumb0) [reply libp2p/HostCapability.h]: It doesn't own, it stores a weak_ptr `Capability::m_session`

Comment (gumb0) [reply libp2p/HostCapability.h]: Yeah I should change the constructor to take weak_ptr, too

Comment (chfast) [reply libp2p/HostCapability.h]: It's not obvious here if `_s` is passed by reference.

Comment (chfast) [reply libp2p/HostCapability.h]: Are you going to change it here?

**Comment by gumb0** [comment]
@chfast Issues addressed.

**Pull request merged**